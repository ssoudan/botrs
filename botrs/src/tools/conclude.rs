use std::cell::RefCell;
use std::fmt::Debug;

use llm_chain::tools::{Describe, Format, Tool, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};

use crate::tools::{TerminalTool, TerminationMessage};

/// A tool that is called to wrap the task.
#[derive(Default)]
pub struct ConcludeTool {
    done: RefCell<Option<ConcludeToolInput>>,
}

impl TerminalTool for ConcludeTool {
    fn is_done(&self) -> bool {
        self.done.borrow().is_some()
    }

    fn take_done(&self) -> Option<TerminationMessage> {
        self.done.borrow_mut().take().map(|input| input.into())
    }
}

/// A tool that is called to wrap the task.
#[derive(Debug, Serialize, Deserialize)]
pub struct ConcludeToolInput {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

impl From<ConcludeToolInput> for TerminationMessage {
    fn from(input: ConcludeToolInput) -> Self {
        Self {
            conclusion: input.conclusion,
            original_question: input.original_question,
        }
    }
}

/// ConcludeToolOutput - empty
#[derive(Serialize, Deserialize)]
pub struct ConcludeToolOutput {}

impl Describe for ConcludeToolInput {
    fn describe() -> Format {
        vec![(
            "conclusion",
            "The final textual answer for this task. No string interpolation supported, only plain text. MANDATORY.",
            )
            .into(),
            (
                "original_question",
                "The original question that was asked to the user. No string interpolation supported, only plain text. MANDATORY.",
            ).into(),]
        .into()
    }
}

impl Describe for ConcludeToolOutput {
    fn describe() -> Format {
        vec![].into()
    }
}

impl ConcludeTool {
    fn invoke_typed(&self, input: ConcludeToolInput) -> Result<ConcludeToolOutput, ToolUseError> {
        if self.done.borrow().is_some() {
            return Err(ToolUseError::ToolInvocationFailed(
                "This task is already done.".to_string(),
            ));
        }

        *self.done.borrow_mut() = Some(input);

        Ok(ConcludeToolOutput {})
    }
}

impl Tool for ConcludeTool {
    fn description(&self) -> ToolDescription {
        ToolDescription::new(
            "Conclude",
            "A tool to conclude a task.",
            "You have to use this to once you have the answer to the task with your conclusion.",
            ConcludeToolInput::describe(),
            ConcludeToolOutput::describe(),
        )
    }

    fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(input)?;
        Ok(serde_yaml::to_value(output)?)
    }
}
