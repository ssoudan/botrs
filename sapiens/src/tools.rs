use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::Arc;

pub use llm_chain::parsing::{find_yaml, ExtractionError};
pub use llm_chain::tools::{Describe, Format, FormatPart, ToolDescription};
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;

/// Error while using a tool
#[derive(Debug, thiserror::Error)]
pub enum ToolUseError {
    /// Tool not found
    #[error("Tool not found: {0}")]
    ToolNotFound(String),
    /// Tool invocation failed
    #[error("Tool invocation failed: {0}")]
    ToolInvocationFailed(String),
    /// Invalid YAML
    #[error("Invalid YAML")]
    InvalidYaml(#[from] serde_yaml::Error),
    /// Invalid input
    #[error("Invalid input: {0}")]
    InvalidInput(#[from] ExtractionError),
}

#[derive(Serialize, Deserialize, Debug)]
struct ToolInvocationInput {
    command: String,
    input: serde_yaml::Value,
    output: Option<serde_yaml::Value>,
}

/// Something meant to become a [`Tool`] - description
pub trait ProtoToolDescribe {
    /// the description of the tool
    fn description(&self) -> ToolDescription;
}

/// Something meant to become a [`Tool`] - invocation
#[async_trait::async_trait]
pub trait ProtoToolInvoke {
    /// Invoke the tool
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

/// A Tool - the most basic kind of tools. See [`AdvancedTool`] and
/// [`TerminalTool`] for more.
#[async_trait::async_trait]
pub trait Tool: Sync + Send {
    /// the description of the tool
    fn description(&self) -> ToolDescription;

    /// Invoke the tool
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

#[async_trait::async_trait]
impl<T: Sync + Send> Tool for T
where
    T: ProtoToolDescribe + ProtoToolInvoke,
{
    fn description(&self) -> ToolDescription {
        self.description()
    }

    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        self.invoke(input).await
    }
}

/// A termination message
///
/// This is the message that is sent to the user when a chain of exchanges
/// terminates.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TerminationMessage {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

/// A [`Tool`] that wraps a chain of exchanges
#[async_trait::async_trait]
pub trait TerminalTool: Tool + Sync + Send {
    /// done flag.
    async fn is_done(&self) -> bool {
        false
    }

    /// Take the done flag.
    async fn take_done(&self) -> Option<TerminationMessage> {
        None
    }
}

/// A [`Tool`]  that can benefit from a [`Toolbox`]
#[async_trait::async_trait]
pub trait AdvancedTool: Tool {
    /// Invoke the tool with a [`Toolbox`]
    async fn invoke_with_toolbox(
        &self,
        toolbox: Toolbox,
        input: serde_yaml::Value,
    ) -> Result<serde_yaml::Value, ToolUseError>;
}

/// Toolbox
///
/// a [`Toolbox`] is a collection of [`Tool`], [`TerminalTool`] and
/// [`AdvancedTool`].
#[derive(Default, Clone)]
pub struct Toolbox {
    /// The terminal tools - the one that can terminate a chain of exchanges
    terminal_tools: Arc<RwLock<HashMap<String, Box<dyn TerminalTool>>>>,

    /// The tools - the other tools
    tools: Arc<RwLock<HashMap<String, Box<dyn Tool>>>>,

    /// The advanced tools - the one that can invoke another tool (not an
    /// advanced one)
    advanced_tools: Arc<RwLock<HashMap<String, Box<dyn AdvancedTool>>>>,
}

impl Debug for Toolbox {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Toolbox").finish()
    }
}

impl Toolbox {
    /// Collect the termination messages
    pub async fn termination_messages(&self) -> Vec<TerminationMessage> {
        let mut messages = Vec::new();

        for tool in self.terminal_tools.read().await.values() {
            if let Some(message) = tool.take_done().await {
                messages.push(message);
            }
        }

        messages
    }

    /// Add a terminal tool
    ///
    /// A [`TerminalTool`] can terminate a chain of exchanges.
    pub async fn add_terminal_tool(&mut self, tool: impl TerminalTool + 'static) {
        let name = tool.description().name;
        self.terminal_tools
            .write()
            .await
            .insert(name, Box::new(tool));
    }

    /// Add a tool
    ///
    /// A [`Tool`] can be invoked by an [`AdvancedTool`].
    pub async fn add_tool(&mut self, tool: impl Tool + 'static) {
        let name = tool.description().name;
        self.tools.write().await.insert(name, Box::new(tool));
    }

    /// Add an advanced tool
    ///
    /// An [`AdvancedTool`] is a [`Tool`] that can invoke another tool.
    pub async fn add_advanced_tool(&mut self, tool: impl AdvancedTool + 'static) {
        let name = tool.description().name;
        self.advanced_tools
            .write()
            .await
            .insert(name, Box::new(tool));
    }

    /// Get the descriptions of the tools
    pub async fn describe(&self) -> HashMap<String, ToolDescription> {
        let mut descriptions = HashMap::new();

        for (name, tool) in self.terminal_tools.read().await.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.tools.read().await.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.advanced_tools.read().await.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        descriptions
    }
}

/// Invoke a [`Tool`] or [`AdvancedTool`]  from a [`Toolbox`]
pub async fn invoke_from_toolbox(
    toolbox: Toolbox,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is an advanced tool
    if let Some(tool) = toolbox.clone().advanced_tools.read().await.get(name) {
        return tool.invoke_with_toolbox(toolbox, input).await;
    }

    // if not, test if the tool is a terminal tool
    {
        let guard = toolbox.terminal_tools.read().await;
        if let Some(tool) = guard.get(name) {
            return tool.invoke(input).await;
        }
    }

    // otherwise, use the normal tool
    let guard = toolbox.tools.read().await;
    let tool = guard
        .get(name)
        .ok_or(ToolUseError::ToolNotFound(name.to_string()))?;

    tool.invoke(input).await
}

/// Invoke a Tool from a [`Toolbox`]
pub async fn invoke_simple_from_toolbox(
    toolbox: Toolbox,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is a terminal tool
    {
        let guard = toolbox.terminal_tools.read().await;
        if let Some(tool) = guard.get(name) {
            return tool.invoke(input).await;
        }
    }

    // the normal tool only
    let guard = toolbox.tools.read().await;
    let tool = guard
        .get(name)
        .ok_or(ToolUseError::ToolNotFound(name.to_string()))?;

    tool.invoke(input).await
}

/// Try to find the tool invocation from the chat message and invoke the
/// corresponding tool.
///
/// If multiple tool invocations are found, only the first one is used.
#[tracing::instrument]
pub async fn invoke_tool(toolbox: Toolbox, data: &str) -> (String, Result<String, ToolUseError>) {
    let tool_invocations = find_yaml::<ToolInvocationInput>(data);

    match tool_invocations {
        Ok(tool_invocations) => {
            if tool_invocations.is_empty() {
                return (
                    "unknown".to_string(),
                    Err(ToolUseError::ToolInvocationFailed(
                        "No Action found".to_string(),
                    )),
                );
            }

            // if any tool_invocations have an 'output' field, we return an error
            for invocation in tool_invocations.iter() {
                if invocation.output.is_some() {
                    return (
                        "unknown".to_string(),
                        Err(ToolUseError::ToolInvocationFailed(
                            "The Action cannot have an `output` field. Only `command` and `input` are allowed.".to_string(),
                        )),
                    );
                }
            }

            // Take the first invocation - the list is reversed
            let invocation_input = &tool_invocations.last().unwrap();

            let tool_name = invocation_input.command.clone();

            let input = invocation_input.input.clone();

            match invoke_from_toolbox(toolbox, &invocation_input.command, input).await {
                Ok(o) => (tool_name, Ok(serde_yaml::to_string(&o).unwrap())),
                Err(e) => (tool_name, Err(e)),
            }
        }
        Err(e) => ("unknown".to_string(), Err(ToolUseError::InvalidInput(e))),
    }
}
