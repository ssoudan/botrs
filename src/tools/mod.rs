/// Bash tool
pub mod bash;
/// Hue tools
pub mod hue;
