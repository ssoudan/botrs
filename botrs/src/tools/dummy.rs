use std::fmt::Debug;

use llm_chain::tools::{Describe, Format, Tool, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};

/// A tool that is called to test stuffs
#[derive(Default)]
pub struct DummyTool {}

/// A tool that is called to test stuffs
#[derive(Debug, Serialize, Deserialize)]
pub struct DummyToolInput {
    /// Well. MANDATORY.
    pub blah: String,
}

/// DummyToolOutput not very significant
#[derive(Serialize, Deserialize)]
pub struct DummyToolOutput {
    /// No much.
    pub something: String,
}

impl Describe for DummyToolInput {
    fn describe() -> Format {
        vec![("blah", "Well. MANDATORY.").into()].into()
    }
}

impl Describe for DummyToolOutput {
    fn describe() -> Format {
        vec![("something", "No much.").into()].into()
    }
}

impl DummyTool {
    fn invoke_typed(&self, input: &DummyToolInput) -> Result<DummyToolOutput, ToolUseError> {
        Ok(DummyToolOutput {
            something: input.blah.clone() + " and something else",
        })
    }
}

impl Tool for DummyTool {
    fn description(&self) -> ToolDescription {
        ToolDescription::new(
            "Dummy",
            "A tool to test stuffs.",
            "Use this to test stuffs.",
            DummyToolInput::describe(),
            DummyToolOutput::describe(),
        )
    }

    fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(&input)?;
        Ok(serde_yaml::to_value(output)?)
    }
}
