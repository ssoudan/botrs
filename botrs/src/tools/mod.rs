use std::collections::HashMap;
use std::rc::Rc;

use llm_chain::tools::{Tool, ToolDescription, ToolUseError};

/// Hue tools
pub mod hue;

/// Tool to conclude a chain
pub mod conclude;

/// Tool to run some (limited) python
pub mod python;

#[cfg(test)]
/// Tool to test stuffs
pub(crate) mod dummy;

/// A termination message
pub struct TerminationMessage {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

/// A tool that wraps a chain of exchanges
pub trait TerminalTool: Tool {
    /// done flag.
    fn is_done(&self) -> bool {
        false
    }

    /// Take the done flag.
    fn take_done(&self) -> Option<TerminationMessage> {
        None
    }
}

/// A tool that can benefit from a toolbox
pub trait AdvancedTool: Tool {
    /// Invoke the tool with a toolbox
    fn invoke_with_toolbox(
        &self,
        toolbox: Rc<Toolbox>,
        input: serde_yaml::Value,
    ) -> Result<serde_yaml::Value, ToolUseError>;
}

/// Toolbox
#[derive(Default)]
pub struct Toolbox {
    /// The terminal tools
    terminal_tools: HashMap<String, Box<dyn TerminalTool>>,

    /// The tools
    tools: HashMap<String, Box<dyn Tool>>,

    /// The advanced tools - the one that can invoke another tool (not an
    /// advanced one)
    advanced_tools: HashMap<String, Box<dyn AdvancedTool>>,
}

impl Toolbox {
    /// Collect the termination messages
    pub fn termination_messages(&self) -> Vec<TerminationMessage> {
        let mut messages = Vec::new();

        for tool in self.terminal_tools.values() {
            if let Some(message) = tool.take_done() {
                messages.push(message);
            }
        }

        messages
    }

    /// Add a terminal tool
    pub fn add_terminal_tool(&mut self, tool: impl TerminalTool + 'static) {
        let name = tool.description().name;
        self.terminal_tools.insert(name, Box::new(tool));
    }

    /// Add a tool
    pub fn add_tool(&mut self, tool: impl Tool + 'static) {
        let name = tool.description().name;
        self.tools.insert(name, Box::new(tool));
    }

    /// Add an advanced tool
    pub fn add_advanced_tool(&mut self, tool: impl AdvancedTool + 'static) {
        let name = tool.description().name;
        self.advanced_tools.insert(name, Box::new(tool));
    }

    /// Get the descriptions
    pub fn describe(&self) -> HashMap<String, ToolDescription> {
        let mut descriptions = HashMap::new();

        for (name, tool) in self.terminal_tools.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.tools.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.advanced_tools.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        descriptions
    }
}

/// Invoke a Tool or AdvancedTool  from a toolbox
pub fn invoke_from_toolbox(
    toolbox: Rc<Toolbox>,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is an advanced tool
    if let Some(tool) = toolbox.clone().advanced_tools.get(name) {
        return tool.invoke_with_toolbox(toolbox, input);
    }

    // if not, test if the tool is a terminal tool
    if let Some(tool) = toolbox.terminal_tools.get(name) {
        return tool.invoke(input);
    }

    // otherwise, use the normal tool
    let tool = toolbox.tools.get(name).ok_or(ToolUseError::ToolNotFound)?;

    tool.invoke(input)
}

/// Invoke a Tool from a toolbox
pub fn invoke_simple_from_toolbox(
    toolbox: Rc<Toolbox>,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is a terminal tool
    if let Some(tool) = toolbox.terminal_tools.get(name) {
        return tool.invoke(input);
    }

    // the normal tool only
    let tool = toolbox.tools.get(name).ok_or(ToolUseError::ToolNotFound)?;

    tool.invoke(input)
}
