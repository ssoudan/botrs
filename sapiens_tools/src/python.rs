use std::rc::Rc;

use convert_case::{Case, Casing};
use pyo3::prelude::*;
use pyo3::types::{IntoPyDict, PyDict, PyFloat, PyList, PyTuple};
use sapiens::tools::{
    invoke_simple_from_toolbox, AdvancedTool, Describe, Format, FormatPart, ProtoToolDescribe,
    ProtoToolInvoke, ToolDescription, ToolUseError, Toolbox,
};
use sapiens_derive::{Describe, ProtoToolDescribe};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;

const MAX_OUTPUT_SIZE: usize = 512;

/// A tool that runs sandboxed Python code. Use this to transform data.
/// - Only stdout and stderr are captured and made available (limited to 512B
/// total).
/// - To use other Tools from here: `input = {...}; output =
/// tools.tool_name(**input); print(output["field_xxx"])`. The `output` is an
/// object.
/// - List available tools with `tools.list()`. `tools` is already
/// imported.
/// - `open`|`exec` are forbidden.
/// - Limited libraries available: urllib3, requests, sympy, numpy,
/// BeautifulSoup4, feedparser.
/// - No PIP.
#[derive(Default, ProtoToolDescribe)]
#[tool(
    name = "SandboxedPython",
    input = "PythonToolInput",
    output = "PythonToolOutput"
)]
pub struct PythonTool {}

/// The input of the Python tool
#[derive(Serialize, Deserialize, Describe)]
pub struct PythonToolInput {
    /// The Python code to run. MANDATORY
    pub code: String,
}

/// The output of the Python tool
#[derive(Serialize, Deserialize, Describe)]
pub struct PythonToolOutput {
    /// The stdout output of the Python code.
    pub stdout: String,
    /// The stderr output of the Python code.
    pub stderr: String,
}

#[pyclass]
#[derive(Default)]
struct Logging {
    output: String,
}

#[pymethods]
impl Logging {
    fn write(&mut self, data: &str) {
        self.output.push_str(data);
    }
}

#[pyclass(unsendable)]
struct ToolsWrapper {
    toolbox: Rc<Toolbox>,
}

impl ToolsWrapper {
    fn new(toolbox: Rc<Toolbox>) -> Self {
        ToolsWrapper { toolbox }
    }
}

#[derive(thiserror::Error, Debug)]
enum PyConversionError {
    #[error("Invalid conversion: {error}")]
    InvalidConversion { error: String },
    #[error("dict key not serializable: {typename}")]
    DictKeyNotSerializable { typename: String },
    #[error("Invalid cast: {typename}")]
    InvalidCast { typename: String },
}

// inspired from https://github.com/mozilla-services/python-canonicaljson-rs/blob/62599b246055a1c8a78e5777acdfe0fd594be3d8/src/lib.rs#L87-L167
fn to_yaml(py: Python, obj: &PyObject) -> Result<Value, PyConversionError> {
    macro_rules! return_cast {
        ($t:ty, $f:expr) => {
            if let Ok(val) = obj.downcast::<$t>(py) {
                return $f(val);
            }
        };
    }

    macro_rules! return_to_value {
        ($t:ty) => {
            if let Ok(val) = obj.extract::<$t>(py) {
                return serde_yaml::to_value(val).map_err(|error| {
                    PyConversionError::InvalidConversion {
                        error: format!("{}", error),
                    }
                });
            }
        };
    }

    if obj.is_none(py) {
        return Ok(Value::Null);
    }

    return_to_value!(String);
    return_to_value!(bool);
    return_to_value!(u64);
    return_to_value!(i64);

    return_cast!(PyDict, |x: &PyDict| {
        let mut map = serde_yaml::Mapping::new();
        for (key_obj, value) in x.iter() {
            let key = if key_obj.is_none() {
                Ok("null".to_string())
            } else if let Ok(val) = key_obj.extract::<bool>() {
                Ok(if val {
                    "true".to_string()
                } else {
                    "false".to_string()
                })
            } else if let Ok(val) = key_obj.str() {
                Ok(val.to_string())
            } else {
                Err(PyConversionError::DictKeyNotSerializable {
                    typename: key_obj
                        .to_object(py)
                        .as_ref(py)
                        .get_type()
                        .name()
                        .map(|x| x.to_string())
                        .unwrap_or_else(|_| "unknown".to_string()),
                })
            };
            map.insert(Value::String(key?), to_yaml(py, &value.to_object(py))?);
        }
        Ok(Value::Mapping(map))
    });

    return_cast!(PyList, |x: &PyList| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyTuple, |x: &PyTuple| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyFloat, |x: &PyFloat| {
        Ok(Value::Number(serde_yaml::Number::from(x.value())))
    });

    // At this point we can't cast it, set up the error object
    Err(PyConversionError::InvalidCast {
        typename: obj
            .as_ref(py)
            .get_type()
            .name()
            .map(|x| x.to_string())
            .unwrap_or_else(|_| "unknown".to_string()),
    })
}

fn value_to_object(val: Value, py: Python<'_>) -> PyObject {
    match val {
        Value::Null => py.None(),
        Value::Bool(x) => x.to_object(py),
        Value::Number(x) => {
            let oi64 = x.as_i64().map(|i| i.to_object(py));
            let ou64 = x.as_u64().map(|i| i.to_object(py));
            let of64 = x.as_f64().map(|i| i.to_object(py));
            oi64.or(ou64).or(of64).expect("number too large")
        }
        Value::String(x) => x.to_object(py),
        Value::Sequence(x) => {
            let inner: Vec<_> = x.into_iter().map(|x| value_to_object(x, py)).collect();
            inner.to_object(py)
        }
        Value::Mapping(x) => {
            let iter = x
                .into_iter()
                .map(|(k, v)| (value_to_object(k, py), value_to_object(v, py)));
            IntoPyDict::into_py_dict(iter, py).into()
        }
        Value::Tagged(_) => panic!("tagged values are not supported"),
    }
}

/// Format of a field in a tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleFormat {
    /// Name of the field
    pub name: String,
    // pub r#type: String,
    /// Description of the field
    pub description: String,
}

impl From<FormatPart> for SimpleFormat {
    fn from(part: FormatPart) -> Self {
        SimpleFormat {
            name: part.key,
            // r#type: part.r#type,
            description: part.purpose,
        }
    }
}

impl ToPyObject for SimpleFormat {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.to_object(py)).unwrap();
        // dict.set_item("type", self.r#type.to_object(py))
        //     .unwrap();
        dict.set_item("description", self.description.to_object(py))
            .unwrap();
        dict.into()
    }
}

/// A simplified version of the tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleToolDescription {
    /// Name of the tool
    pub name: String,
    /// Description of the tool
    pub description: String,
    /// Context utilization of the tool
    pub description_context: String,
    /// Input format of the tool
    pub input_format: Vec<SimpleFormat>,
    /// Output format of the tool
    pub output_format: Vec<SimpleFormat>,
}

impl From<ToolDescription> for SimpleToolDescription {
    fn from(desc: ToolDescription) -> Self {
        let input_format = desc
            .input_format
            .parts
            .into_iter()
            .map(|x| x.into())
            .collect();
        let output_format = desc
            .output_format
            .parts
            .into_iter()
            .map(|x| x.into())
            .collect();

        SimpleToolDescription {
            name: desc.name,
            description: desc.description,
            description_context: desc.description_context,
            input_format,
            output_format,
        }
    }
}

impl ToPyObject for SimpleToolDescription {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.clone()).unwrap();
        dict.set_item("description", self.description.clone())
            .unwrap();
        dict.set_item("description_context", self.description_context.clone())
            .unwrap();
        dict.set_item("input_format", self.input_format.clone())
            .unwrap();
        dict.set_item("output_format", self.output_format.clone())
            .unwrap();
        dict.into()
    }
}

#[pymethods]
impl ToolsWrapper {
    // list all tools
    #[pyo3(signature = ())]
    fn list(&self, py: Python<'_>) -> PyResult<PyObject> {
        let tools = self.toolbox.describe();
        let tools = tools
            .into_values()
            .map(SimpleToolDescription::from)
            .collect::<Vec<_>>();
        let tools = tools.to_object(py);
        Ok(tools)
    }

    // invoke a tool
    #[pyo3(signature = (tool_name, input))]
    fn invoke(
        &self,
        py: Python<'_>,
        tool_name: &str,
        input: Option<&PyDict>,
    ) -> PyResult<PyObject> {
        // convert PyDict to a serde_yaml::Value
        let input = if let Some(input) = input {
            let input: PyObject = input.into();

            to_yaml(py, &input).map_err(|e| {
                pyo3::exceptions::PyException::new_err(format!("Invalid input: {}", e))
            })?
        } else {
            Value::default()
        };

        // println!("invoking tool {} with input {:?}", tool_name, input);

        let output =
            invoke_simple_from_toolbox(self.toolbox.clone(), tool_name, input).map_err(|e| {
                pyo3::exceptions::PyException::new_err(format!("Tool invocation failed: {}", e))
            })?;

        let output = value_to_object(output, py);

        Ok(output)
    }
}

impl PythonTool {
    fn invoke_typed(
        &self,
        toolbox: Option<Rc<Toolbox>>,
        input: &PythonToolInput,
    ) -> Result<PythonToolOutput, ToolUseError> {
        let mut code = input.code.clone();

        // check for forbidden keywords - with capture
        let re = regex::Regex::new(r"(exec|pip)").unwrap();
        if let Some(caps) = re.captures(&code) {
            return Err(ToolUseError::ToolInvocationFailed(format!(
                "Python code contains forbidden keywords such as {}",
                caps.get(0).unwrap().as_str()
            )));
        }

        let tools = toolbox.map(ToolsWrapper::new);

        // dynamically add functions to a `tools` module
        if let Some(tools) = &tools {
            let mut tool_class_code = String::new();

            tool_class_code.push_str("class Tools:\n");

            tool_class_code.push_str("    def __init__(self, toolbox):\n");
            tool_class_code.push_str("        self.toolbox = toolbox\n");

            for (name, description) in tools.toolbox.as_ref().describe() {
                let inputs_parts = description.input_format.parts;
                let inputs = inputs_parts
                    .iter()
                    .map(|f| f.key.clone())
                    .collect::<Vec<_>>()
                    .join(", ");
                let inputs = if inputs.is_empty() {
                    "".to_string()
                } else {
                    format!("(self, {})", inputs)
                };

                let dict = inputs_parts
                    .iter()
                    .map(|f| {
                        let name = &f.key;
                        format!("\"{}\": {}", name, name)
                    })
                    .collect::<Vec<_>>()
                    .join(", ");

                // in snake case
                tool_class_code.push_str(&format!(
                    "    def {}{}:\n        return self.toolbox.invoke(\"{}\", {{{}}})\n",
                    name.to_case(Case::Snake),
                    inputs,
                    name,
                    dict
                ));

                // in Pascal case
                tool_class_code.push_str(&format!(
                    "    def {}{}:\n        return self.toolbox.invoke(\"{}\", {{{}}})\n",
                    name.to_case(Case::Pascal),
                    inputs,
                    name,
                    dict
                ));

                // FUTURE(ssoudan) set input_format and output_format
            }

            // add list function
            tool_class_code.push_str("    def list(self):\n");
            tool_class_code.push_str("        return self.toolbox.list()\n");

            tool_class_code.push_str("tools = Tools(toolbox)\n");

            // prepend the tool class code to the user code
            code = format!("{}\n{}", tool_class_code, code);
        }

        // print!("{}", code);

        let res: PyResult<(String, String)> = Python::with_gil(|py| {
            // println!("Python version: {}", py.version());

            let globals = if let Some(tools) = tools {
                let tools_cell = PyCell::new(py, tools)?;
                [("toolbox", tools_cell)].into_py_dict(py)
            } else {
                PyDict::new(py)
            };

            // capture stdout and stderr
            let sys = py.import("sys")?;

            let stdout = Logging::default();
            let py_stdout_cell = PyCell::new(py, stdout)?;
            let py_stdout = py_stdout_cell.borrow_mut();
            sys.setattr("stdout", py_stdout.into_py(py))?;

            let stderr = Logging::default();
            let py_stderr_cell = PyCell::new(py, stderr)?;
            let py_stderr = py_stderr_cell.borrow_mut();
            sys.setattr("stderr", py_stderr.into_py(py))?;

            // FUTURE(ssoudan) pass something in

            // run code
            Python::run(py, &code, globals.into(), None)?;

            // NOFUTURE(ssoudan) get something out

            let stdout = py_stdout_cell.borrow().output.clone();
            let stderr = py_stderr_cell.borrow().output.clone();

            Ok((stdout, stderr))
        });

        let (stdout, stderr) = res.map_err(|e| {
            ToolUseError::ToolInvocationFailed(format!("Python code execution failed: {}", e))
        })?;

        Ok(PythonToolOutput { stdout, stderr })
    }
}

// impl ProtoToolDescribe for PythonTool {
//     fn description(&self) -> ToolDescription {
//         ToolDescription::new(
//             "SandboxedPython",
//             &format!("A tool that executes sandboxed Python code. Only stdout
// and stderr are captured and made available (limited to {}B total). ",
// MAX_OUTPUT_SIZE),             r#"Use this to transform data. To use other
// Tools from here: `input = {...}; output = tools.tool_name(**input);
// print(output["field_xxx"])`. List available tools with `tools.list()` -
// `tools` is already imported. The `output` is an object. open|exec are
// forbidden. Limited libraries available: urllib3, requests, sympy, numpy,
// BeautifulSoup4, feedparser. No PIP."#,
// PythonToolInput::describe(),             PythonToolOutput::describe(),
//         )
//     }
// }

impl ProtoToolInvoke for PythonTool {
    fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(None, &input)?;

        // check the size of the output (stdout and stderr)
        let l = output.stdout.len() + output.stderr.len();
        if l > MAX_OUTPUT_SIZE {
            return Err(ToolUseError::ToolInvocationFailed(format!(
                "Python code produced too much output on stdout and stderr combined ({} bytes) - max is {}",
                l,
                MAX_OUTPUT_SIZE
            )));
        }

        Ok(serde_yaml::to_value(output)?)
    }
}

impl AdvancedTool for PythonTool {
    fn invoke_with_toolbox(
        &self,
        toolbox: Rc<Toolbox>,
        input: Value,
    ) -> Result<Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(Some(toolbox), &input)?;
        Ok(serde_yaml::to_value(output)?)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use pyo3::indoc::indoc;

    use super::*;
    use crate::dummy::DummyTool;

    #[test]
    fn test_python_tool() {
        let tool = PythonTool::default();
        let input = PythonToolInput {
            code: indoc! {
            r#"print('hello')
               t = toolbox.list()
               print("tools=", t)
               
               d = tools.dummy(blah="ahah")
               print("dummy=", d)
               
               "#}
            .to_string(),
        };
        let mut toolbox = Toolbox::default();
        toolbox.add_tool(DummyTool::default());
        let toolbox = Rc::new(toolbox);

        let output = tool.invoke_typed(Some(toolbox), &input).unwrap();
        assert_snapshot!(output.stdout);
        assert_snapshot!(output.stderr);
    }
}
