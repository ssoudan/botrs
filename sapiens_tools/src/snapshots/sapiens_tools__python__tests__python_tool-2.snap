---
source: sapiens_tools/src/python.rs
expression: output.stderr
---

